accepted/0000-default-command.md

```diff
@@ -0,0 +1,173 @@
+
+## Detailed Explanation
+
+- Add a new config key, `default-command`.  This value must be a known npm
+  command (no aliases or abbreviations), and will have the default value of
+  `'help'`.
+- When the first positional argument does not match a known npm subcommand,
+  it will invoke the `default-command` option with the remainder of the
+  arguments.
+- When the first argument _does_ match a known npm subcommand, then the
+  `default-command` config value is not invoked.  (For example, `npm
+  --default-command=run-script install foo` will still install `foo` if
+  there is an `install` script in `package.json`.)
+- If a new command is added in the future, and this means that current
+  users of `npm foo` being an alias for `npm run-script foo`, then that's
+  an acceptable disruption.  New commands are added very rarely, and only
+  after a careful evaluation of existing userland packages.  But, this is a
+  power-user convenience feature, and we do not expect _many_ of our users
+  will take advantage of it (though we expect those that use it will be
+  happy to have it, and will understand the risks).
+- The `default-command` must only be a single command name.  It does not
+  support setting configurations or specifying additional arguments.  For
+  example, `npm config set default-command 'install --global'` would not be
+  allowed.
+
+## Rationale and Alternatives
+
+`npm help` is a good default for most users, especially new users.
+However, power users and those coming from Yarn may prefer to run their
+build scripts with `npm build` rather than `npm run build`.
+
+The rationale for being willing to clobber convenience use cases (eg,
+scripts named "install", etc.) when a new command is added is twofold.
+First, we are somewhat following in Yarn's footsteps, and Yarn only
+auto-runs scripts when the command is not known.  Second, we have to
+reserve the right to continue to expand the functionality of the CLI, or
+else development of npm is unreasonably limited.
+
+That said, it _does_ impose a non-zero cost in adding new commands to the
+CLI, since users _may_ be relying on a `--default-command` config to handle
+that term in some other way.
+
+### Alternative 1 - run-script _and then_ help
+
+One alternative would be to search for a match among known npm subcommands,
+then `run-script` options, and only _then_ fall back to searching help.
+
+This would be least surprising to current Yarn users, still pretty helpful,
+and not require any opt-in.  However, it is not extensible to other command
+use cases, and still imposes the cost of adding new commands.
+
+### Alternative 2 - just make `help` know about scripts
+
+Another alternative would be to make the current default `help-search`
+behavior look at the `"scripts"` field in the current project's
+`package.json` file.  If the term matches there, it could suggest that the
+user might be thinking of the `run-script` command.
```

Comment:
this seems like a good improvement regardless, and would help the default/common case.

Reply:
Yeah, it's somewhat orthogonal to this RFC, and something we could just do regardless.